### ISSUE TITLE
Spike: New and improved Moped Test deploy pipeline

### ISSUE THREAD

**Issue by mateoclarke** [cityofaustin/atd-data-tech#9261]
_Let's time box this Spike to ~4-8 hours._

Requirements for an improved Moped Test Environment solution:
- Able to deploy multiple test environments, not just one dedicated instance like we have with Heroku. (We can compromise on this in the short term if the overhead is too extreme)
- Deploy trigger from a UI that is accessible to non-devs (ex: Github Actions).
- Deploy to AWS (so that test is as close to production as possible)
- (Optional) Gets copy of current Moped production data instead of seed data.

Example user story:
1. User triggers GH Action inputting a branch for the deployment
2. New Hasura cluster builds on AWS as a brand new instance with a production replica DB
3. User gets URL link to the new frontend
4. Process for spinning down an instance no longer in use.

Other notes/questions:
- Are AWS CloudFormation Templates are useful solution here? Ansible? Terraform? etc
- Is there a suggested solution/preference within the open source Hasura community?
- How can we make it easy/not dangerous to view active test instances and spin one down (without, .e.g, accidentally terminating prod)

**Comment by mddilley** [comment]
Looks like Hasura has [preview apps](https://hasura.io/docs/latest/graphql/cloud/preview-apps/) figured out in their paid Hasura Cloud service. Maybe, there are some clues on how they do it in their docs/[code](https://github.com/hasura/hasura-cloud-preview-apps) that we could use in our self-hosted stack? Some of the steps sound similar to what we all talked about in the meeting today so that is good to know. 
1. GH Action to start the building preview app (autobuild for them but we can set up like the current action to specify branch)
2. Create an ephemeral DB on an existing Postgres server
3. Delete the preview database

I did a tiny bit of looking into CloudFormation, but I think if we go the route of using an existing RDS instance then it wouldn't be useful since CloudFormation can only provision new instances and not databases within an instance.

BTW is Moped already using CloudFormation stacks? I didn't specifically ask, and I was never super familiar with the deployment but I didn't think so. I still need to poke around the AWS console. 

CloudFormation initially sound like a cool way to compartmentalize the test instances since you can easily delete a stack when its no longer needed. But, if we can use existing resources, that sounds like a great way to go!

**Comment by mateoclarke** [comment]
Terraform module worth reviewing: https://github.com/Rayraegah/terraform-aws-hasura

**Comment by mddilley** [comment]
Met with Frank today, and, after hearing Frank's idea about using a Prefect flow to orchestrate, we decided that is a good path forward since we already are using Prefect on the team. Creating a flow will also help us visualize and break down the steps needed to manage test instances.

**Initials tasks**
- Find where the test instance will differ from staging/prod
  - Use Netlify? Probably so but there are alternatives if easier
  - Use existing staging RDS for ephemeral DBs (instead of dedicated instances like prod has)
- Stub out the steps of the flow to create pockets of work to be filled in (and Frank brought up the advantage of being able to get multiple developers working at once and getting exposure to the test instance deployment)

@frankhereford I know that you are going to add notes from your brainstorming 🧠 so feel free to add/edit this as needed. 🙏

**Comment by frankhereford** [comment]
## Summary

🙏 Thanks @mddilley for a great sync just now. We had a chance to brainstorm for a little while and start laying out our plan of execution to get this stood up as a MVP to bring back to the group for thoughts and assessment. As we discussed, Mike is  going to serve as the director for this team effort, and he'll be laying out our first tasks soon. Stay tuned 📺 

##  Brainstorming  ideas

### 🧠 🌧️

_Very_ broadly, we want a script that can orchestrate the creation, configuration and destruction of AWS / Netlify resources. A vision of this script could have the following properties:

* This script is actually a python flow. I'm going to say flow for the rest of this comment, but you can use it as a script / CLI  if you want, but it's out-of-the-box integrated into prefect and controlled by prefect.
  * Prefect offers a lot to this -- it can provide the state / working-memory we need to do things like poll clusters that are running to see if they are still being used and take actions if they are not. It can capture logs of deployments and share them for debugging. It's a champ at handling inter-dependent tasks, such as standing up a cluster of multiple technologies and services. 
*  The script runs on an agent on `atd-data03` so we can safely make accessible to it all of the little file and secret-like-things that need to be used but don't lend themselves to prefect's KV store mechanism. Think about the `authorized_keys` files that would let us SSH to a deployment if needed, for example. 
* The script uses Ansible or similar to give us a high-level interface to operate with instead of having to deal directly with the AWS / netlify API. A very first step is decide which to use. There are great options.
* The script has 4 core capabilities
  * Create a cluster for a feature branch
  * Destroy a cluster
  * List / identify a cluster that is running
  * Protect a cluster from being auto-harvested by the garbage collector
* The script would be able to create a cluster comprised of:
  * A replicated database on the RDS server from live moped (or optionally seed data)
  * A EC2 instance or ECS instance which runs our version of `graphql-engine`
  * A Netlify deployment of the build
  * A DNS entry via Route 53 that is the endpoint for the graphql-engine (Stretch goal, IP hostname for `graphql-engine` backend is fine. Netlify also does fine if you don't mind using their domain (like we do now.)
  * Another DNS entry (R53 again) for the netlify if needed. (Another stretch goal, this is just fanciness so we have a custom domain instead of netlify in the address bar.)
 * A final thought; writing this as a flow gives us a great framework for collaboration. The flow boilerplate is like pretty much nothing, and we're out of the gate stubbing in functions named after all the steps we spoke about above. We use the prefect functional API or the imperative API to link them into a DAG, and we'll be getting to visualize the interdependencies of a deployment action's moving parts for free. It's almost self-organizing; by thinking like a prefect flow, we start out in the "let's do a bunch of interdependent tiny steps", which is the name of the game for the problem we're trying to solve. And, it will let us show "how" we're going to do something to the group before we actually do the heavy lifting of composing the functions inside the stubs.

**Comment by mddilley** [comment]
Synced with @frankhereford again today, and we went through [the skeleton flow](https://github.com/cityofaustin/atd-prefect/blob/9261-md-moped-test-instance-flow/flows/moped/moped_editor_test_instance_deploy.py) and prioritized two sets of tasks to focus on as next steps:

1. Creating and deleting a database for the preview app on an isolated and small RDS instance (me)
2. Creating and deleting the ECS cluster for the Hasura GraphQL engine (Frank)

Other tasks in the skeleton are lower in priority or are stretch goals. The primary focus is achieve the same level of functionality of the current moped-test instance and Netlify deploy previews. The nice-to-have's can fill in as we develop this further. We also discussed sticking to `boto3` for AWS until a tool like Ansible makes sense to incorporate.
